enhancements/network/sdn-live-migration.md

```diff
@@ -0,0 +1,442 @@
+     according to the HostSubnet CR of this node.
+     
+   CNO wait until OVN-K is deployed in the cluster Set the
+   `LiveMigrationTargetCNIReady` condition to `TRUE` in the status of the
+   network.config CR.
+
+6. CNO updates the `.status.migration.networkType` of the network.config CR.
+   This change will trigger MCO to apply a new MachineConfig to each MCP:
+
+   - MCO will render a new MachineConfig for each MCP.
+   - MCO will cordon, drain and reboot the node.
+   - When the node is up, the `br-ex` bridge will be created. It means the node
+     is ready for OVN Kubernetes. So the local zone OVN Kubernetes components
+     can be started. The ovnkube-node container wrapper script will also add the
+     node annotation `k8s.ovn.org/node-subnets`, according to the HostSubnet CR
+     of this node. On the local node, OVN Kubernetes will add all SDN nodes as
+     hybrid overlay nodes to the local zone and all OVN nodes as OVN-IC remote
+     zones. On a remote OVN node, OVN Kubernetes will remove a hybrid overlay
+     node and add an OVN-IC remote zone. Pods will be recreated on the node
+     using ovn-kubernetes as the default CNI plugin.
+   - MCO will uncordon the node. New pods can be scheduled to this node.
+
+   The above process will be repeated for each node until all the nodes have
+   been applied to the new MachineConfig and converted to OVN Kubernetes.
+
+7. CNO watches the MCPs, wait until the new MCs is applied to all the nodes. Set
+   the `LiveMigrationTargetMachineConfigApplied` condition to `TRUE` in the
+   status of the network.config CR.
+
+8. CNO removes the `spec.migration` field of network.operator CR. It will
+   trigger the CNO to:
+
+   - delete the openshift-sdn DaemonSets and the related resources
+     (CustomResources, ConfigMaps, etc.).
+   - redeploy ovn-kubernetes in "normal" mode (no migration mode script, hybrid
+     overlay disabled, etc.).
+   - remove the migration-related labels and node annotations from the nodes.
+
+9. CNO wait until OVN-K is redeployed Set the `LiveMigrationOriginalCNIPurged`
+   condition to `TRUE` in the status of the network.config CR. Then set the
+   `LiveMigrationProgressing` condition to `FALSE` with the reason
+   `LiveMigrationCompleted` in the status of the network.config CR. This
+   indicates that the migration is completed.
+
+### API
+
+For `Network.config.openshift.io` CRD, we will use the CR annotation
+`network.openshift.io/live-migration:` to gate the modification to
+`spec.networkType` field. If the annotation absents (default), CNO will block
+the change to take effect. If it exists, CNO takes it as the trigger of starting
+SDN live migration.
+
+The following conditions will be added to the `.status` of the
+`Network.config.openshift.io` CR:
+
+- LiveMigrationProgressing
+- LiveMigrationRoutableMTUApplied
+- LiveMigrationTargetCNIReady
+- LiveMigrationTargetMachineConfigApplied
+- LiveMigrationOriginalCNIPurged
```

Comment:
I am unsure about the need and the meaning of some these status entries. Are they merely informative, or are they needed for something else, like enabling specific troubleshooting procedures that will be documented? Or do they follow a design pattern?

I am concerned that the value of these statuses can be confusing if there several migration procedures are attempted. Should they be reset to false when a new migration or rollback is started?

I will leave it up to @cblecker and @jewzaam as I am sure they are more knowledgeable than me on how statuses can be leveraged.

Reply:
These conditions will be used by CNO to determine the current state of live migration. These conditions will be set to ` FALSE` when a live migration is triggered and reset to `UNKNOW` when a live migration is completed. Also, they will be used for troubleshooting.

Reply:
Isn't all this worth mentioning here or is it a common design pattern with status conditions? You mentioned to me there might be documentation around this. Could you refer me to it? Thanks!

Reply:
Yes, I can add this here. Here's the document about conditions https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties